### PR TITLE
[WIN32SS][USER32] Refactor USER32 callbacks

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Refactoring kernel-to-`USER32` callback code.
Localize USER32 callback entry definitions.
JIRA issue: N/A

## Proposed changes

- Add header file `win32ss/include/u32cb.h` and use it to define the `USER32` callback entries.

### CODE REVIEW

**Comment by HBelusca** [comment]
> Add {{U32CB}}: comments for USER32 callback functions

Why?

File: win32ss/user/user32/windows/cursoricon.c
Comment (HBelusca) [inline win32ss/user/user32/windows/cursoricon.c:45]: Put this description just after the callback ID comment instead?

File: win32ss/include/callback.h
Comment (HBelusca) [inline win32ss/include/callback.h:6]:
```suggestion
typedef enum _USER32_CALLBACK
```

**Comment by katahiromz** [comment]
> > Add {{U32CB}}: comments for USER32 callback functions
> 
> Why? (and especially, why this `{{U32CB}}:` syntax?)

Machine friendly code (I will feed them to bots).

**Comment by JoachimHenze** [comment]
I do dislike this PR, especially I do see absolutely no relation of it to CORE-19485 'JPEG/PNG/GIF/TIFF wallpaper support'.
I am not convinced why we would need to do that.

File: win32ss/user/user32/windows/cursoricon.c
Comment (katahiromz) [reply win32ss/user/user32/windows/cursoricon.c]: Deleted comments.

**Comment by katahiromz** [comment]
Deleted `{{U32CB}}:` comments.
Added wallpaper support code.

**Comment by katahiromz** [comment]
AFTER:
after
Successful.

File: win32ss/include/callback.h
Comment (JoachimHenze) [inline win32ss/include/callback.h:15]: Please use an 8.3 filename for this new file, e.g. "u32cb.h" or "user32cb.h"

Comment (katahiromz) [reply win32ss/include/callback.h]: Fixed.

File: win32ss/user/user32/misc/loadpic.c
Comment (katahiromz) [inline win32ss/user/user32/misc/loadpic.c:126]: @Doug-Lyons You can insert the Vista icon related code here in the future. Just a note.

**Review by JoachimHenze — APPROVED** [review]
Approved, but please wait for a 2nd approval before doing the merge.

**Comment by HBelusca** [comment]
> Add User32CallLoadPictureFromKernel and co_ClientLoadPicture functions to load a non-BMP picture.

Does this exist on Windows (2003 or Vista+)? If not, you already know your PR is a hack.

**Comment by HBelusca** [comment]
This PR is invalid because of many reasons:
- The most blatant one is _**Layering violation**_ (read https://web.archive.org/web/20140420035454/http://blogs.msdn.com/b/larryosterman/archive/2005/08/23/455193.aspx  for what I mean), where you make user32 call a higher-level functionality: Gdiplus (Gdi+) to deal with the non-bitmap wallpaper;
- JPEG/PNG/etc. wallpaper support already exist, and is done for example via the control panel applet that calls Gdi+ to convert such a wallpaper file into bitmap, as it is on Windows. On Windows 7 the mspaint uses instead a COM call to a procedure of the CLSID_ActiveDesktop interface, but the idea isthe same.
- When choosing a jpeg/etc. wallpaper from mspaint, mspaint already knows how to deal with such image format (via Cdi+), and will save the wallpaper as a bitmap file that is set as wallpaper. Again, just check what happens on Windows.

**Comment by HBelusca** [comment]
> > Does this exist on Windows (2003 or Vista+)?
> 
> (image)

This is on Windows 2003 or Vista+?

**Comment by katahiromz** [comment]
> This is on Windows 2003 or Vista+?

Win2k3.

**Review by HBelusca — CHANGES_REQUESTED** [review]
This PR is invalid because of many reasons:
- The most blatant one is _**Layering violation**_ (read https://web.archive.org/web/20140420035454/http://blogs.msdn.com/b/larryosterman/archive/2005/08/23/455193.aspx  for what I mean), where you make user32 call a higher-level functionality: Gdiplus (Gdi+) to deal with the non-bitmap wallpaper;
- JPEG/PNG/etc. wallpaper support already exist, and is done for example via the control panel applet that calls Gdi+ to convert such a wallpaper file into bitmap, as it is on Windows. On Windows 7 the mspaint uses instead a COM call to a procedure of the CLSID_ActiveDesktop interface, but the idea isthe same.
- When choosing a jpeg/etc. wallpaper from mspaint, mspaint already knows how to deal with such image format (via Gdi+), and will save the wallpaper as a bitmap file that is set as wallpaper. Again, just check what happens on Windows.

**Comment by katahiromz** [comment]
Closed due to wrong implementation and saving review time. I will re-implement the right thing.

**Comment by HBelusca** [comment]
@katahiromz You could always keep what your original branch and PR was doing, namely putting the callback definitions in a file that is included; this would have been OK.

**Comment by katahiromz** [comment]
Deleted wallpaper code.
Now callback refactoring only.

**Pull request merged**
